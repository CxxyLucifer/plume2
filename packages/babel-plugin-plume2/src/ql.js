/**
 * 判断当前的callexpress是不是QL
 * @param {*} path 
 */
export const isQLCallExpress = path => {
  const { node } = path;
  return node.callee.name === 'QL';
};

/**
 * 是不是可用的QL的语法
 * @param {*} path 
 */
export const isValidQLSyntax = path => {
  const { node } = path;
};
