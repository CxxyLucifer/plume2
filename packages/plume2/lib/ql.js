"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var uuid = 0;
var QueryLang = /** @class */ (function () {
    function QueryLang(name, lang) {
        this._id = ++uuid;
        this._name = name;
        this._lang = lang;
    }
    QueryLang.prototype.id = function () {
        return this._id;
    };
    QueryLang.prototype.name = function () {
        return this._name;
    };
    QueryLang.prototype.lang = function () {
        return this._lang;
    };
    QueryLang.prototype.setLang = function (lang) {
        this._lang = lang;
        return this;
    };
    return QueryLang;
}());
exports.QueryLang = QueryLang;
function QL(name, lang) {
    return new QueryLang(name, lang);
}
exports.QL = QL;
