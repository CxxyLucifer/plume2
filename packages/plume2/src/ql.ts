export type Lang = Array<any>;

let uuid = 0;

export class QueryLang {
  constructor(name: string, lang: Lang) {
    this._id = ++uuid;
    this._name = name;
    this._lang = lang;
  }

  private _id: number;
  private _name: string;
  private _lang: Lang;

  id() {
    return this._id;
  }

  name() {
    return this._name;
  }

  lang() {
    return this._lang;
  }

  setLang(lang: Array<any>) {
    this._lang = lang;
    return this;
  }
}

export function QL(name: string, lang: Lang) {
  return new QueryLang(name, lang);
}
