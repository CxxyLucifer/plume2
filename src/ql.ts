import { isArray, isString } from './type';

export type Lang = Array<any>;

let uuid = 0;

export class QueryLang {
  private _id: number;
  private _name: string;
  private _lang: Lang;

  constructor(name: string, lang: Lang) {
    this._id = ++uuid;
    this._name = name;
    this._lang = lang;
  }

  id() {
    return this._id;
  }

  name() {
    return this._name;
  }

  lang() {
    return this._lang;
  }

  setLang(lang: Array<any>) {
    this._lang = lang;
    return this;
  }
}

export class DynamicQueryLang {
  private _ctx: Object;
  private _name: string;
  private _lang: Array<any>;

  constructor(name: string, lang: Array<any>) {
    this._ctx = {};
    this._name = name;
    this._lang = lang;
  }

  //parse DQL's lang
  //return QL's lang
  analyserLang(dLang: Array<any>): Array<any> {
    const lang = [];

    for (let i = 0, len = dLang.length; i < len; i++) {
      let path = dLang[i];

      if (isString(path)) {
        lang[i] = path[0] === '$' ? this._ctx[path.substring(1)] : path;
      } else if (isArray(path)) {
        lang[i] = [];
        for (let j = 0, len = path.length; j < len; j++) {
          let field = dLang[i][j];
          lang[i][j] =
            isString(field) && field[0] === '$'
              ? this._ctx[field.substring(1)]
              : field;
        }
      } else if (path instanceof DynamicQueryLang) {
        //each time, we create a new QueryLang
        //in this case, we don't need any cache
        lang[i] = new QueryLang(
          `${path._name}2QL`,
          this.analyserLang(path._lang)
        );
      } else {
        lang[i] = path;
      }
    }

    return lang;
  }

  withContext(ctx: Object) {
    this._ctx = ctx;
    return this;
  }

  name() {
    return this._name;
  }

  lang() {
    return this._lang;
  }
}

function isDQLLang(lang: Array<any>): boolean {
  //why len - 1? üòù
  for (let i = 0, len = lang.length; i < len - 1; i++) {
    const path = lang[i];
    //Â¶ÇÊûúÂΩìÂâçÁöÑË∑ØÂæÑÊòØÂ≠óÁ¨¶‰∏≤ÔºåÁõ¥Êé•Âà§Êñ≠ÊòØ‰∏çÊòØÂåÖÂê´'$'
    //Â¶ÇÊûúÂΩìÂâçÁöÑË∑ØÂæÑÊòØÊï∞ÁªÑÔºåjoin‰πãÂêéÂà§Êñ≠ÊòØ‰∏çÊòØÂåÖÂê´'$'
    //Â¶ÇÊûúÂΩìÂâçÁöÑË∑ØÂæÑÊòØDynamicLang Áõ¥Êé•ËøîÂõûtrue
    if (isString(path) && path[0] === '$') {
      return true;
    } else if (isArray(path) && path.join('').indexOf('$') != -1) {
      return true;
    } else if (path instanceof DynamicQueryLang) {
      return true;
    }
  }

  return false;
}

//expose
export function QL(name: string, lang: Lang) {
  return isDQLLang(lang)
    ? new DynamicQueryLang(name, lang)
    : new QueryLang(name, lang);
}
