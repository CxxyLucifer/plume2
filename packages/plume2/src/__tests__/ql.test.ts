import { Actor, Store } from '../index';
import { QL, QueryLang } from '../ql';

class LoadingActor extends Actor {
  defaultState() {
    return { loading: true };
  }
}

class TodoActor extends Actor {
  defaultState() {
    return { todo: [{ id: 1, text: 'hello plume', done: false }] };
  }
}

class AppStore extends Store {
  bindActor() {
    return [new LoadingActor(), new TodoActor()];
  }
}

describe('ql test suite', () => {
  it('init ql state', () => {
    const helloQL = QL('helloQL', [
      'loading',
      /**
       *
       */
      loading => loading
    ]) as QueryLang;

    expect(helloQL.id()).toEqual(1);
    expect(helloQL.name()).toEqual('helloQL');
    expect(helloQL.lang()).toMatchSnapshot();
  });

  it('test bigQuery', () => {
    const todoQL = QL('todoQL', [
      'loading',
      ['todo', 0, 'text'],
      (loading, text) => ({
        loading,
        text
      })
    ]) as QueryLang;

    const store = new AppStore({});
    const todo = store.bigQuery(todoQL);

    expect(todo).toEqual({
      loading: true,
      text: 'hello plume'
    });
  });

  it('test complex bigQuery', () => {
    const loadingQL = QL('loadingQL', [
      'loading',
      loading => loading
    ]) as QueryLang;

    const todoQL = QL('todoQL', [
      loadingQL,
      ['todo', 0, 'text'],
      (loading, text) => ({
        loading,
        text
      })
    ]) as QueryLang;

    const store = new AppStore({});
    const loading = store.bigQuery(loadingQL);
    expect(loading).toEqual(true);

    const todo = store.bigQuery(todoQL);
    expect(todo).toEqual({ loading: true, text: 'hello plume' });

    //from cache
    const todoCache = store.bigQuery(todoQL);
    expect(todoCache).toEqual({ loading: true, text: 'hello plume' });
  });

  it('when path value is undefined', () => {
    const testQL = QL('testQL', ['a', a => 10]);
    const store = new AppStore();
    const result = store.bigQuery(testQL as QueryLang);

    expect(result).toEqual(10);
  });
});
