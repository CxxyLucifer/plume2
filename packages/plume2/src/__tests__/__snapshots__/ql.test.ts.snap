// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ql test suite init ql state 1`] = `
Array [
  "loading",
  [Function],
]
`;
